document_wopi/models/document_wopi_access_token.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+# © 2017 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import time
+import uuid
+import urllib
+import urlparse
+from datetime import timedelta
+from openerp import api, fields, models
+from openerp.exceptions import AccessDenied
+
+
+class DocumentWopiAccessToken(models.Model):
+    _name = 'document.wopi.access.token'
+    _description = 'Access token for WOPI requests'
+
+    user_id = fields.Many2one(
+        'res.users', 'User', required=True, ondelete='cascade',
+    )
+    attachment_id = fields.Many2one(
+        'ir.attachment', 'Attachment', required=True, ondelete='cascade',
+    )
+    valid_until = fields.Datetime()
+    token = fields.Char(required=True)
+    token_ttl = fields.Integer(compute='_compute_token_ttl')
+    action_url = fields.Char(compute='_compute_action_url')
+
+    @api.model
+    def _get_token(self, attachment):
+        """Create or return a token for some attachment for the current user"""
+        existing = self.search([
+            ('user_id', '=', self.env.user.id),
+            ('attachment_id', '=', attachment.id),
+            '|',
+            ('valid_until', '=', False),
+            ('valid_until', '>=', fields.Datetime.now()),
+        ], limit=1)
+        if existing:
+            return existing
+        return self.create({
+            'user_id': self.env.user.id,
+            'attachment_id': attachment.id,
+            'valid_until': fields.Datetime.to_string(
+                fields.Datetime.from_string(fields.Datetime.now()) +
+                timedelta(self.env['ir.config_parameter'].get_param(
+                    'document_wopi.token_validity', 36000,
+                ))
+            ),
+            # TODO: is this random enough or do we need some secret generator?
+            'token': str(uuid.uuid4()),
+        })
+
+    @api.multi
+    def _compute_token_ttl(self):
+        for this in self:
+            if not this.valid_until:
+                continue
+            this.token_ttl = int(time.mktime(
+                fields.Datetime.from_string(this.valid_until).timetuple()
+            ))
+
+    @api.multi
+    def _compute_action_url(self):
+        # TODO: this needs to look up the value from the last call to
+        # /hosting/discovery. The current value probably is installation
+        # specific
+        for this in self:
+            this.action_url = self.env['document.wopi']._get_url(
+                'loleaflet/cda55f3/loleaflet.html'
```

Comment:
yes, this is not production ready. WOPI does service discovery, we need to save those URLs and use them. This is specific for the current docker version of libreoffice online.